File: app/src/main/java/com/example/bottlr/SharedUtils.java
package com.example.bottlr;

import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Environment;
import android.provider.MediaStore;
import android.widget.Toast;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.core.content.FileProvider;
import java.util.List;
import java.util.ArrayList;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Files;

public class SharedUtils {

    //region App-Wide Methods

    //region Safe Read

    // Safe value read.
    public static String readValueSafe(BufferedReader br) throws IOException {
        String line = br.readLine();
        return (line != null && line.contains(": ")) ? line.split(": ", 2)[1] : "";
    }

    //endregion

    //region Parse Bottle from Save
    public static Bottle parseBottle(File file) {
        try (BufferedReader br = new BufferedReader(new FileReader(file))) {

            String name = readValueSafe(br);

            if (name.isEmpty()) {
                return null;
                // Make sure the bottle is a valid entry
            }
            String distillery = readValueSafe(br);
            String type = readValueSafe(br);
            String abv = readValueSafe(br);
            String age = readValueSafe(br);
            String notes = readValueSafe(br);
            String region = readValueSafe(br);
            String keywords = readValueSafe(br);
            String rating = readValueSafe(br);
            String photoUriString = readValueSafe(br);
            Uri photoUri = (!photoUriString.equals("No photo") && !photoUriString.isEmpty()) ? Uri.parse(photoUriString) : null;

            return new Bottle(name, distillery, type, abv, age, photoUri, notes, region, keywords, rating);

            // Exception handling
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static Cocktail parseCocktail(File file) {
        try (BufferedReader br = new BufferedReader(new FileReader(file))) {

            String name = readValueSafe(br);

            if (name.isEmpty()) {
                return null;
                // Make sure the bottle is a valid entry
            }
            String base = readValueSafe(br);
            String mixer = readValueSafe(br);
            String juice = readValueSafe(br);
            String liqueur = readValueSafe(br);
            String garnish = readValueSafe(br);
            String extra = readValueSafe(br);
            String notes = readValueSafe(br);
            String keywords = readValueSafe(br);
            String rating = readValueSafe(br);
            String photoUriString = readValueSafe(br);
            Uri photoUri = (!photoUriString.equals("No photo") && !photoUriString.isEmpty()) ? Uri.parse(photoUriString) : null;

            return new Cocktail(name, base, mixer, juice, liqueur, garnish, extra, photoUri, notes, keywords, rating);

            // Exception handling
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    //endregion

    //region Shopping Query Builder

    // Search query string builder
    public static String queryBuilder(Bottle toBuy) {
        StringBuilder queryBuilder = new StringBuilder();

        // Append other fields only if they are not empty
        if (toBuy.getDistillery() != null && !toBuy.getDistillery().isEmpty()) {
            queryBuilder.append(" ").append(toBuy.getDistillery()).append(" ");
        }

        // Always include name
        queryBuilder.append(toBuy.getName());

        if (toBuy.getAge() != null && !toBuy.getAge().isEmpty()) {
            queryBuilder.append(" ").append(toBuy.getAge()).append(" Year");
        }
        if (toBuy.getRegion() != null && !toBuy.getRegion().isEmpty()) {
            queryBuilder.append(" ").append(toBuy.getRegion());
        }
        if (toBuy.getType() != null && !toBuy.getType().isEmpty()) {
            queryBuilder.append(" ").append(toBuy.getType());
        }
        return queryBuilder.toString();
    }

    //endregion

    //region Share Functionalities

    //region shareBottleInfo

    // Shares bottle information
    public static void shareBottleInfo(Bottle bottle, Context context) {
        if (bottle != null) {
            String shareText = createShareText(bottle);
            Uri imageUri = (bottle.getPhotoUri() != null && !bottle.getPhotoUri().toString().equals("No photo"))
                    ? cacheImage(bottle.getPhotoUri(), context)
                    : null;
            shareBottleContent(shareText, imageUri, context);
        }
    }

    public static void shareCocktailInfo(Cocktail cocktail, Context context) {
        if (cocktail != null) {
            String shareText = createShareTextCocktail(cocktail);
            Uri imageUri = (cocktail.getPhotoUri() != null && !cocktail.getPhotoUri().toString().equals("No photo"))
                    ? cacheImage(cocktail.getPhotoUri(), context)
                    : null;
            shareBottleContent(shareText, imageUri, context); //should still work
        }
    }

    //endregion

    //region cacheImage

    // Saves the bottle image to the cache for use in sharing bottle details.
    public static Uri cacheImage(Uri imageUri, Context context) {
        try {
            // Changed to use try-with-resources to automatically close the InputStream
            try (InputStream inputStream = context.getContentResolver().openInputStream(imageUri)) {
                String fileName = "cached_bottle_image_" + System.currentTimeMillis() + ".png";
                File cacheFile = new File(context.getFilesDir(), fileName);
                try (OutputStream outputStream = Files.newOutputStream(cacheFile.toPath())) {
                    byte[] buffer = new byte[4096]; // Adjust this buffer size if needed
                    int bytesRead;
                    while (true) {
                        assert inputStream != null;
                        if ((bytesRead = inputStream.read(buffer)) == -1) break;
                        outputStream.write(buffer, 0, bytesRead);
                    }
                }
                return FileProvider.getUriForFile(context, "com.example.bottlr.fileprovider", cacheFile);
            }
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    //endregion

    //region shareBottleContent

    // For sharing Bottle details
    public static void shareBottleContent(String text, @Nullable Uri imageUri, Context context) {
        Intent shareIntent = new Intent();
        shareIntent.setAction(Intent.ACTION_SEND);
        shareIntent.putExtra(Intent.EXTRA_TEXT, text);
        shareIntent.setType("text/plain");

        if (imageUri != null) {
            shareIntent.putExtra(Intent.EXTRA_STREAM, imageUri);
            shareIntent.setType("image/*");
            shareIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        }

        context.startActivity(Intent.createChooser(shareIntent, "Share Bottle Info"));
    }

    //endregion

    //region Bottle ShareText Builder

    // Constructs a share post using pertinent bottle details, if available.
    public static String createShareText(Bottle bottle) {
        StringBuilder shareText = new StringBuilder("Here's what I'm drinking:\n\n");

        // Always include the bottle name, since it's always there
        shareText.append(bottle.getName());
        // Include the distillery statement if one exists
        if (bottle.getDistillery() != null && !bottle.getDistillery().isEmpty()) {
            shareText.append(", by ").append(bottle.getDistillery());
        }
        shareText.append("\n");

        // Check for Age, Region, and Type fields
        boolean hasAge = bottle.getAge() != null && !bottle.getAge().isEmpty();
        boolean hasRegion = bottle.getRegion() != null && !bottle.getRegion().isEmpty();
        boolean hasType = bottle.getType() != null && !bottle.getType().isEmpty();

        if (hasAge || hasRegion || hasType) {
            // Get to work constructing the post with available data
            if (hasAge) {
                // Add age statement, if applicable
                shareText.append(bottle.getAge()).append("-Year ");
            }
            if (hasRegion) {
                // Add region, if applicable
                shareText.append(bottle.getRegion()).append(" ");
            }
            if (hasType) {
                // Add type, if applicable
                shareText.append(bottle.getType());
            }
            shareText.append("\n");
        }

        // Adding notes, if applicable
        if (bottle.getNotes() != null && !bottle.getNotes().isEmpty()) {
            shareText.append("\nMy Thoughts:\n").append(bottle.getNotes());
        }

        // Finalize string. Should look like this:
        //
        // Here's what I'm drinking:
        //
        // The Peat Seat, by Speynnigan's Wake
        // 5-Year Speyside Scotch
        //
        // My Thoughts:
        // Peaty, smoky, nice notes of caramel.
        //
        // Any missing details should be omitted in formatting.

        return shareText.toString();
    }

    public static String createShareTextCocktail(Cocktail cocktail) { //copied from bottle
        StringBuilder shareText = new StringBuilder("I'm drinking a ");
        shareText.append(cocktail.getName());
        shareText.append(".\n\n");
        shareText.append("Thanks Bottlr!");

        // Finalize string. Should look like this:
        //
        // I'm drinking a Gin and Tonic.
        //
        // Thanks Bottlr!

        return shareText.toString();
    }

    //endregion

    //endregion

    //region Deletion Handling

    // Confirmation Popup
    public static void showDeleteConfirm(final Bottle bottle, Context context) {
        new AlertDialog.Builder(context)
                .setTitle("Delete Bottle")
                .setMessage("Are you sure you want to delete this bottle?")
                .setPositiveButton(android.R.string.yes, (dialog, which) -> deleteBottle(bottle, context))
                .setNegativeButton(android.R.string.no, null)
                .setIcon(android.R.drawable.ic_dialog_alert)
                .show();
    }

    public static void showDeleteConfirmCocktail(final Cocktail cocktail, Context context) {
        new AlertDialog.Builder(context)
                .setTitle("Delete Cocktail")
                .setMessage("Are you sure you want to delete this cocktail?")
                .setPositiveButton(android.R.string.yes, (dialog, which) -> deleteBottle(cocktail, context))
                .setNegativeButton(android.R.string.no, null)
                .setIcon(android.R.drawable.ic_dialog_alert)
                .show();
    }

    // Deletion Code
    public static void deleteBottle(Bottle bottle, Context context) {
        String filename = "bottle_" + bottle.getName() + ".txt";
        File file = new File(context.getFilesDir(), filename);
        boolean deleted = file.delete();
        if (deleted) {
            Toast.makeText(context, "Bottle Deleted.", Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(context, "Failed To Delete Bottle.", Toast.LENGTH_SHORT).show();
        }
    }

    public static void deleteCocktail(Cocktail cocktail, Context context) {
        String filename = "cocktail_" + cocktail.getName() + ".txt";
        File file = new File(context.getFilesDir(), filename);
        boolean deleted = file.delete();
        if (deleted) {
            Toast.makeText(context, "Cocktail Deleted.", Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(context, "Failed To Delete Cocktail.", Toast.LENGTH_SHORT).show();
        }
    }
    //endregion

    //region Bottle List Loading
    public static List<Bottle> loadBottles(Context context) {
        List<Bottle> bottles = new ArrayList<>();
        File directory = context.getFilesDir();
        File[] files = directory.listFiles();

        assert files != null;
        for (File file : files) {
            if (file.isFile() && file.getName().startsWith("bottle_")) {
                Bottle bottle = parseBottle(file);
                if (bottle != null) {
                    bottles.add(bottle); // Add the bottle to the list
                }
            }
        }
        return bottles;
    }

    public static List<Cocktail> loadCocktails(Context context) {
        List<Cocktail> cocktails = new ArrayList<>();
        File directory = context.getFilesDir();
        File[] files = directory.listFiles();

        assert files != null;
        for (File file : files) {
            if (file.isFile() && file.getName().startsWith("cocktail_")) {
                Cocktail cocktail = parseCocktail(file);
                if (cocktail != null) {
                    cocktails.add(cocktail); // Add the bottle to the list
                }
            }
        }
        return cocktails;
    }
    //endregion

    //region Save Image to Gallery
    public static void saveImageToGallery(Context context, Bottle bottle) {
        // Convert the Uri to a Bitmap
        try {
            Bitmap bitmap = MediaStore.Images.Media.getBitmap(context.getContentResolver(), bottle.getPhotoUri());

            // Prepare the values for the new image
            ContentValues values = new ContentValues();
            values.put(MediaStore.Images.Media.DISPLAY_NAME, bottle.getName() + "_BottlrSavedImage.jpg");
            values.put(MediaStore.Images.Media.MIME_TYPE, "image/jpeg");
            values.put(MediaStore.Images.Media.RELATIVE_PATH, Environment.DIRECTORY_PICTURES);

            // Add the new image to the MediaStore.Images collection
            Uri imageUri = context.getContentResolver().insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, values);
            if (imageUri != null) {
                try (OutputStream outputStream = context.getContentResolver().openOutputStream(imageUri)) {
                    assert outputStream != null;
                    bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream);
                } catch (IOException e) {
                    e.printStackTrace();
                    Toast.makeText(context, "Failed To Save Image", Toast.LENGTH_SHORT).show();
                }
            }
            // TODO: Get this notification working
            //else {
            //    Toast.makeText(context, "No Image To Save", Toast.LENGTH_SHORT).show();
            //}

            // Show a toast message
            Toast.makeText(context, "Saved As " + bottle.getName() + "_BottlrSavedImage.jpg", Toast.LENGTH_SHORT).show();
        } catch (IOException e) {
            e.printStackTrace();
            Toast.makeText(context, "Failed To Save Image", Toast.LENGTH_SHORT).show();
        }
    }
    public static void saveImageToGalleryCocktail(Context context, Cocktail cocktail) {
        // Convert the Uri to a Bitmap
        try {
            Bitmap bitmap = MediaStore.Images.Media.getBitmap(context.getContentResolver(), cocktail.getPhotoUri());

            // Prepare the values for the new image
            ContentValues values = new ContentValues();
            values.put(MediaStore.Images.Media.DISPLAY_NAME, cocktail.getName() + "_BottlrSavedImage.jpg");
            values.put(MediaStore.Images.Media.MIME_TYPE, "image/jpeg");
            values.put(MediaStore.Images.Media.RELATIVE_PATH, Environment.DIRECTORY_PICTURES);

            // Add the new image to the MediaStore.Images collection
            Uri imageUri = context.getContentResolver().insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, values);
            if (imageUri != null) {
                try (OutputStream outputStream = context.getContentResolver().openOutputStream(imageUri)) {
                    assert outputStream != null;
                    bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream);
                } catch (IOException e) {
                    e.printStackTrace();
                    Toast.makeText(context, "Failed To Save Image", Toast.LENGTH_SHORT).show();
                }
            }
            // TODO: Get this notification working
            //else {
            //    Toast.makeText(context, "No Image To Save", Toast.LENGTH_SHORT).show();
            //}

            // Show a toast message
            Toast.makeText(context, "Saved As " + cocktail.getName() + "_BottlrSavedImage.jpg", Toast.LENGTH_SHORT).show();
        } catch (IOException e) {
            e.printStackTrace();
            Toast.makeText(context, "Failed To Save Image", Toast.LENGTH_SHORT).show();
        }
    }
    //endregion

    //endregion
}

File: app/src/main/kotlin/com/bottlr/app/SharedUtils.kt
package com.bottlr.app

import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Environment
import android.provider.MediaStore
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.core.content.FileProvider
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.nio.file.Files

// Safely reads a value from a BufferedReader.
fun readValueSafe(br: BufferedReader): String {
    val line = br.readLine()
    return if (line != null && line.contains(": ")) line.split(": ", limit = 2)[1] else ""
}

// Parses a Bottle from a saved file.
fun parseBottle(file: File): Bottle? {
    return try {
        BufferedReader(FileReader(file)).use { br ->
            val name = readValueSafe(br)
            if (name.isEmpty()) return null

            val distillery = readValueSafe(br)
            val type = readValueSafe(br)
            val abv = readValueSafe(br)
            val age = readValueSafe(br)
            val notes = readValueSafe(br)
            val region = readValueSafe(br)
            val keywords = readValueSafe(br)
            val rating = readValueSafe(br)
            val photoUriString = readValueSafe(br)
            val photoUri = if (photoUriString.isNotBlank() && photoUriString != "No photo") Uri.parse(photoUriString) else null

            Bottle(name, distillery, type, abv, age, photoUri, notes, region, keywords, rating)
        }
    } catch (e: IOException) {
        e.printStackTrace()
        null
    }
}

// Parses a Cocktail from a saved file.
fun parseCocktail(file: File): Cocktail? {
    return try {
        BufferedReader(FileReader(file)).use { br ->
            val name = readValueSafe(br)
            if (name.isEmpty()) return null

            val base = readValueSafe(br)
            val mixer = readValueSafe(br)
            val juice = readValueSafe(br)
            val liqueur = readValueSafe(br)
            val garnish = readValueSafe(br)
            val extra = readValueSafe(br)
            val notes = readValueSafe(br)
            val keywords = readValueSafe(br)
            val rating = readValueSafe(br)
            val photoUriString = readValueSafe(br)
            val photoUri = if (photoUriString.isNotBlank() && photoUriString != "No photo") Uri.parse(photoUriString) else null

            Cocktail(name, base, mixer, juice, liqueur, garnish, extra, photoUri, notes, keywords, rating)
        }
    } catch (e: IOException) {
        e.printStackTrace()
        null
    }
}

// Builds a shopping query string for a Bottle.
fun queryBuilder(toBuy: Bottle): String {
    return buildString {
        toBuy.distillery.takeIf { !it.isNullOrEmpty() }?.let { append(" $it ") }
        append(toBuy.name)
        toBuy.age.takeIf { !it.isNullOrEmpty() }?.let { append(" $it Year") }
        toBuy.region.takeIf { !it.isNullOrEmpty() }?.let { append(" $it") }
        toBuy.type.takeIf { !it.isNullOrEmpty() }?.let { append(" $it") }
    }
}

// Shares information about a Bottle.
fun shareBottleInfo(bottle: Bottle?, context: Context) {
    bottle ?: return
    val shareText = createShareText(bottle)
    val imageUri = bottle.photoUri?.takeIf { it.toString().isNotBlank() && it.toString() != "No photo" }?.let { cacheImage(it, context) }
    shareBottleContent(shareText, imageUri, context)
}

// Shares information about a Cocktail.
fun shareCocktailInfo(cocktail: Cocktail?, context: Context) {
    cocktail ?: return
    val shareText = createShareTextCocktail(cocktail)
    val imageUri = cocktail.photoUri?.takeIf { it.toString().isNotBlank() && it.toString() != "No photo" }?.let { cacheImage(it, context) }
    shareBottleContent(shareText, imageUri, context) // Reusing the same function should be fine
}

// Caches an image for sharing.
fun cacheImage(imageUri: Uri, context: Context): Uri? {
    return try {
        context.contentResolver.openInputStream(imageUri).use { inputStream ->
            val fileName = "cached_bottle_image_${System.currentTimeMillis()}.png"
            val cacheFile = File(context.filesDir, fileName)
            Files.newOutputStream(cacheFile.toPath()).use { outputStream ->
                inputStream?.copyTo(outputStream)
            }
            FileProvider.getUriForFile(context, "com.bottlr.app.fileprovider", cacheFile)
        }
    } catch (e: IOException) {
        e.printStackTrace()
        null
    }
}

// Shares bottle content (text and optional image).
fun shareBottleContent(text: String, imageUri: Uri?, context: Context) {
    val shareIntent = Intent().apply {
        action = Intent.ACTION_SEND
        putExtra(Intent.EXTRA_TEXT, text)
        type = "text/plain"

        imageUri?.let {
            putExtra(Intent.EXTRA_STREAM, it)
            type = "image/*"
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
    }
    context.startActivity(Intent.createChooser(shareIntent, "Share Bottle Info"))
}

// Creates the text for sharing a Bottle.
fun createShareText(bottle: Bottle): String {
    return buildString {
        append("Here's what I'm drinking:\n\n")
        append(bottle.name)
        bottle.distillery.takeIf { !it.isNullOrEmpty() }?.let { append(", by $it") }
        append("\n")

        val hasAge = !bottle.age.isNullOrEmpty()
        val hasRegion = !bottle.region.isNullOrEmpty()
        val hasType = !bottle.type.isNullOrEmpty()

        if (hasAge || hasRegion || hasType) {
            if (hasAge) append("${bottle.age}-Year ")
            if (hasRegion) append("${bottle.region} ")
            if (hasType) append(bottle.type)
            append("\n")
        }

        bottle.notes.takeIf { !it.isNullOrEmpty() }?.let { append("\nMy Thoughts:\n$it") }
    }
}

// Creates the text for sharing a Cocktail.
fun createShareTextCocktail(cocktail: Cocktail): String {
    return "I'm drinking a ${cocktail.name}.\n\nThanks Bottlr!"
}

// Shows a confirmation dialog for deleting a Bottle.
fun showBottleDeleteConfirm(bottle: Bottle, context: Context) {
    AlertDialog.Builder(context)
        .setTitle("Delete Bottle")
        .setMessage("Are you sure you want to delete this bottle?")
        .setPositiveButton(android.R.string.yes) { _, _ -> deleteBottle(bottle, context) }
        .setNegativeButton(android.R.string.no, null)
        .setIcon(android.R.drawable.ic_dialog_alert)
        .show()
}

// Shows a confirmation dialog for deleting a Cocktail.
fun showDeleteConfirmCocktail(cocktail: Cocktail, context: Context) {
    AlertDialog.Builder(context)
        .setTitle("Delete Cocktail")
        .setMessage("Are you sure you want to delete this cocktail?")
        .setPositiveButton(android.R.string.yes) { _, _ -> deleteCocktail(cocktail, context) }
        .setNegativeButton(android.R.string.no, null)
        .setIcon(android.R.drawable.ic_dialog_alert)
        .show()
}

// Deletes a Bottle file.
fun deleteBottle(bottle: Bottle, context: Context) {
    val filename = "bottle_${bottle.name}.txt"
    val file = File(context.filesDir, filename)
    if (file.delete()) {
        Toast.makeText(context, "Bottle Deleted.", Toast.LENGTH_SHORT).show()
    } else {
        Toast.makeText(context, "Failed To Delete Bottle.", Toast.LENGTH_SHORT).show()
    }
}

// Deletes a Cocktail file.
fun deleteCocktail(cocktail: Cocktail, context: Context) {
    val filename = "cocktail_${cocktail.name}.txt"
    val file = File(context.filesDir, filename)
    if (file.delete()) {
        Toast.makeText(context, "Cocktail Deleted.", Toast.LENGTH_SHORT).show()
    } else {
        Toast.makeText(context, "Failed To Delete Cocktail.", Toast.LENGTH_SHORT).show()
    }
}

// Loads all Bottles from storage.
fun loadBottles(context: Context): MutableList<Bottle> {
    val bottles = mutableListOf<Bottle>()
    val directory = context.filesDir
    directory.listFiles()?.forEach { file ->
        if (file.isFile && file.name.startsWith("bottle_")) {
            parseBottle(file)?.let { bottles.add(it) }
        }
    }
    return bottles
}

// Loads all Cocktails from storage.
fun loadCocktails(context: Context): MutableList<Cocktail> {
    val cocktails = mutableListOf<Cocktail>()
    val directory = context.filesDir
    directory.listFiles()?.forEach { file ->
        if (file.isFile && file.name.startsWith("cocktail_")) {
            parseCocktail(file)?.let { cocktails.add(it) }
        }
    }
    return cocktails
}

// Loads all Locations from storage.
fun loadLocations(context: Context): MutableList<Location> {
    val locations = mutableListOf<Location>()
    val directory = context.filesDir
    directory.listFiles()?.forEach { file ->
        if (file.isFile && file.name.startsWith("location_")) {
            parseLocation(file)?.let {
                locations.add(it)
                Log.d("SharedUtils", "Location with name ${it.name} loaded successfully.")
            }
        }
    }
    return locations
}

// Parses a Location from a saved file.
fun parseLocation(file: File): Location? {
    return try {
        BufferedReader(FileReader(file)).use { br ->
            val timeDateAdded = readValueSafe(br)
            if (timeDateAdded.isEmpty()) return null

            val gpsCoordinates = readValueSafe(br)
            val name = readValueSafe(br)

            Log.d("SharedUtils", "Location with name $name parsed successfully.")
            Location(timeDateAdded, gpsCoordinates, name)
        }
    } catch (e: IOException) {
        e.printStackTrace()
        null
    }
}

// Saves an image to the device's gallery.
fun saveImageToGallery(context: Context, bottle: Bottle) {
    try {
        val bitmap = MediaStore.Images.Media.getBitmap(context.contentResolver, bottle.photoUri)
        val values = ContentValues().apply {
            put(MediaStore.Images.Media.DISPLAY_NAME, "${bottle.name}_BottlrSavedImage.jpg")
            put(MediaStore.Images.Media.MIME_TYPE, "image/jpeg")
            put(MediaStore.Images.Media.RELATIVE_PATH, Environment.DIRECTORY_PICTURES)
        }

        val imageUri = context.contentResolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, values)
        imageUri?.let {
            context.contentResolver.openOutputStream(it)?.use { outputStream ->
                bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream)
            }
        }

        Toast.makeText(context, "Saved As ${bottle.name}_BottlrSavedImage.jpg", Toast.LENGTH_SHORT).show()
    } catch (e: IOException) {
        e.printStackTrace()
        Toast.makeText(context, "Failed To Save Image", Toast.LENGTH_SHORT).show()
    }
}

fun saveImageToGalleryCocktail(context: Context, cocktail: Cocktail) {
    try {
        val bitmap = MediaStore.Images.Media.getBitmap(context.contentResolver, cocktail.photoUri)
        val values = ContentValues().apply {
            put(MediaStore.Images.Media.DISPLAY_NAME, "${cocktail.name}_BottlrSavedImage.jpg")
            put(MediaStore.Images.Media.MIME_TYPE, "image/jpeg")
            put(MediaStore.Images.Media.RELATIVE_PATH, Environment.DIRECTORY_PICTURES)
        }

        val imageUri = context.contentResolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, values)
        imageUri?.let {
            context.contentResolver.openOutputStream(it)?.use { outputStream ->
                bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream)
            }
        }

        Toast.makeText(context, "Saved As ${cocktail.name}_BottlrSavedImage.jpg", Toast.LENGTH_SHORT).show()
    } catch (e: IOException) {
        e.printStackTrace()
        Toast.makeText(context, "Failed To Save Image", Toast.LENGTH_SHORT).show()
    }
}

// Shows a dialog with options for a Location.
fun showLocationDialog(location: Location, context: Context) {
    AlertDialog.Builder(context)
        .setTitle("Location Options")
        .setItems(arrayOf("View", "Delete", "Close")) { dialog, which ->
            when (which) {
                0 -> { // View
                    val url = "https://www.google.com/maps?tbm=map&q=${Uri.encode(location.gpsCoordinates)}"
                    val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
                    context.startActivity(intent)
                }
                1 -> showLocationDeleteConfirm(location, context) // Delete
                2 -> dialog.dismiss() // Close
            }
        }
        .create()
        .show()
}

// Shows a confirmation dialog for deleting a Location.
fun showLocationDeleteConfirm(location: Location, context: Context) {
    AlertDialog.Builder(context)
        .setTitle("Delete Location")
        .setMessage("Are you sure you want to delete this location?")
        .setPositiveButton(android.R.string.yes) { _, _ -> deleteLocation(location, context) }
        .setNegativeButton(android.R.string.no, null)
        .setIcon(android.R.drawable.ic_dialog_alert)
        .show()
}

// Deletes a Location file.
fun deleteLocation(location: Location, context: Context) {
    val filename = "location_${location.name}.txt"
    val file = File(context.filesDir, filename)
    if (file.delete()) {
        Toast.makeText(context, "Location Deleted.", Toast.LENGTH_SHORT).show()
    } else {
        Toast.makeText(context, "Failed To Delete Location.", Toast.LENGTH_SHORT).show()
    }
}
